src/components/Widget.tsx

```diff
@@ -114,10 +114,34 @@ export default function Widget(props: PropsWithChildren<WidgetProps>) {
     }
     return props.locale ?? DEFAULT_LOCALE
   }, [props.locale])
-  const jsonRpcEndpoint = useMemo(
-    () => props.jsonRpcEndpoint ?? JSON_RPC_FALLBACK_ENDPOINTS[SupportedChainId.MAINNET]?.[0],
-    [props.jsonRpcEndpoint]
-  )
+  const jsonRpcEndpointIsDict =
+    typeof props.jsonRpcEndpoint !== 'string' && !JsonRpcProvider.isProvider(props.jsonRpcEndpoint)
+  const defaultChainId = useMemo(() => {
+    if (!props.defaultChainId) return 1
+    if (!ALL_SUPPORTED_CHAIN_IDS.includes(props.defaultChainId)) {
+      console.warn(`Unsupported chainId: ${props.defaultChainId}. Falling back to 1 (Ethereum Mainnet).`)
+      return 1
+    }
+    if (props.jsonRpcEndpoint && jsonRpcEndpointIsDict) {
+      if (!Object.keys(props.jsonRpcEndpoint).includes(`${props.defaultChainId}`)) {
+        console.warn(
+          `Did not provide a jsonRpcEndpoint for default chainId: ${props.defaultChainId}. Falling back to 1 (Ethereum Mainnet).`
+        )
+        return 1
+      }
+    }
+    return props.defaultChainId
+  }, [props.defaultChainId])
+  const jsonRpcEndpoint: string | JsonRpcProvider | { [chainId: number]: string[] } = useMemo(() => {
+    if (!props.jsonRpcEndpoint) return JSON_RPC_FALLBACK_ENDPOINTS
+    if (jsonRpcEndpointIsDict && !Object.keys(props.jsonRpcEndpoint).includes(`${defaultChainId}`)) {
+      console.warn(
+        `Did not provide a jsonRpcEndpoint for default chainId: ${defaultChainId}. Falling back to free public endpoints.`
+      )
+      return JSON_RPC_FALLBACK_ENDPOINTS
+    }
+    return props.jsonRpcEndpoint
+  }, [props.jsonRpcEndpoint, defaultChainId])
```

Comment:
There's lots of nested boolean statements here; it'd be smart to clean it up by assigning compound statements to descriptive variable names. I also see a couple spots where you could simplify this flow.

i.e., 
```
    if (props.jsonRpcEndpoint && jsonRpcEndpointIsDict) {
      if (!Object.keys(props.jsonRpcEndpoint).includes(`${props.defaultChainId}`)) {
...
      }
    }
```

could be 
```
    if (props.jsonRpcEndpoint && jsonRpcEndpointIsDict && !Object.keys(props.jsonRpcEndpoint).includes(`${props.defaultChainId}`)) {
...
    }
```

I also think it would make sense to include the truthy/falsy test of `jsonRpcEndpoint` in the `jsonRpcEndpointIsDict` statement